clang/lib/Analysis/UnsafeBufferUsage.cpp

```diff
@@ -282,8 +282,8 @@ isInUnspecifiedPointerContext(internal::Matcher<Stmt> InnerMatcher) {
   //    (i.e., computing the distance between two pointers); or ...
 
   auto CallArgMatcher =
-      callExpr(forEachArgumentWithParam(InnerMatcher,
-                  hasPointerType() /* array also decays to pointer type*/),
+      callExpr(forEachArgumentWithParamType(InnerMatcher,
+                  isAnyPointer() /* array also decays to pointer type*/),
```

Comment:
Original code uses `hasPointerType()` which is inconvenient here because it includes the `hasType()` part, but another thing it was doing was `hasCanonicalType()` which is now missing. Are you sure it's not necessary in this case?

Reply:
Removing `hasCanonicalType()` is intentional but not well communicated by the commit message - I'll fix that.

There's 2 different things going on and I probably could've separated this tiny change into two even-more-tiny changes.
1. Support for calls on function pointers - that's mostly the `forEachArgumentWithParam` -> `forEachArgumentWithParamType`.
2. Suport for const size arrays - that's removing the `hasCanonicalType()` to allow for array to pointer decay (which the comment in the snippet suggests we already expected).

Reply:
Yeah but I mean, does the new code handle typedefs correctly? That's probably why `hasCanonicalType()` was there in the first place, to handle things like typedefs.

Reply:
Yes, type sugar is fine, we have about 20 tests cases for just `typedef`.

From the architectural standpoint this is a definition of DRE context, it shouldn't have used `hasCanonicalType` in the first place and the comment `/* array also decays to pointer type*/` suggests it wasn't the intention.

Handling or not handling typedefs is the responsibility of `fixVariable` function and perhaps individual FixableGadgets1 `getFixits()` method.

Reply:
Hmm looks like parameter types are always canonical. Probably because we need them to be the same for overloading purposes even if they're typedef'ed differently. The new code appears to be entirely equivalent to `hasCanonicalType(pointerType())` or just `pointerType()`.

Side note, `isAnyPointer()` only adds ObjC pointer types to the mix. We don't really care about ObjC pointers because you can't do pointer arithmetic on them. So might as well use `pointerType()`.

> Handling or not handling typedefs is the responsibility of `fixVariable` function and perhaps individual FixableGadgets' `getFixits()` method.

In many cases you need to actively skip typedefs in order to get the gadget to match at all. But it looks like, indeed, it doesn't matter in this case.

Reply:
Summary of an offline discussion: Actually, the above is wrong because we're talking about the parameter type, not argument type but it also voids the concern.